pkg/services/ngalert/api/tooling/definitions/alertmanager_validation.go

```diff
@@ -0,0 +1,67 @@
+package definitions
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/common/model"
+)
+
+// Validate normalizes a Route r, and returns errors if r is invalid.
+func (r *Route) Validate() error {
+	r.GroupBy = nil
+	r.GroupByAll = false
```

Comment:
These two lines are the only change from the original. These are added to make the normalization idempotent - without this, validating a valid object multiple times would eventually result in an error.

This is because the object is designed to have values loaded into `GroupByStr`, and then the normalization process computes other fields based on the value of this. There was a bug where it never reset the computed fields, but this was masked because all this code was embedded in serialization methods.

This makes the validate work correctly, **provided that the caller uses the object as intended and loads data into GroupByStr**. This expectation was present before this change. Redesigning this object to make its API less of a foot-gun would be nice, but is out of scope for this PR, since it focuses purely on making the extracted validation work within the existing use of the object. So, I went with this as a compromise.

Reply:
Seems sensible to me 👍